eva/executor/hash_join_executor.py

```diff
@@ -0,0 +1,60 @@
+# coding=utf-8
+# Copyright 2018-2020 EVA
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Iterator
+
+from eva.models.storage.batch import Batch
+from eva.executor.abstract_executor import AbstractExecutor
+from eva.planner.hash_join_probe_plan import HashJoinProbePlan
+
+
+class HashJoinExecutor(AbstractExecutor):
+
+    def __init__(self, node: HashJoinProbePlan):
+        super().__init__(node)
+        self.predicate = node.join_predicate
+        self.join_type = node.join_type
+        self.probe_keys = node.probe_keys
+        self.join_project = node.join_project
+
+    def validate(self):
+        pass
+
+    def exec(self, *args, **kwargs) -> Iterator[Batch]:
+
+        build_table = self.children[0]
+        probe_table = self.children[1]
+        hash_keys = [key.col_alias for key in self.probe_keys]
+        for build_batch in build_table.exec():
+            for probe_batch in probe_table.exec():
+                probe_batch.frames.index = probe_batch.frames[
+                    hash_keys].apply(
+                    lambda x: hash(tuple(x)), axis=1)
+                join_batch = probe_batch.frames.merge(build_batch.frames,
+                                                      left_index=True,
+                                                      right_index=True,
+                                                      how='inner')
+                join_batch.reset_index(drop=True, inplace=True)
+                join_batch = Batch(join_batch)
+                if not join_batch.empty() and self.predicate:
+                    outcomes = self.predicate.evaluate(join_batch).frames
+                    filtered_df = join_batch.frames[(outcomes > 0).to_numpy()]
+                    join_batch = Batch(filtered_df.reset_index(drop=True))
+                # Then do project
+                if not join_batch.empty() and self.join_project:
+                    batches = [expr.evaluate(join_batch)
+                               for expr in self.join_project]
+                    join_batch = Batch.merge_column_wise(batches)
+
```

Comment:
What is the benefit of having the selection and projection in the join? Instead of say a seqscan's executor above the Join (better modularity). If we later want to make optimizations to the selection and projection's executor , we don't need to change at multiple places.

Reply:
Seq scan operator?

Reply:
I'm not sure if I understand this completely, but I believe you added the predicate and project executor and not using them here?

Reply:
Yes, 
```mermaid
flowchart TD;
   subgraph ide2 [option two]
   PROJECTION --- SELECTION
   SELECTION --- B(JOIN)
   end
   subgraph ide1 [option one]
   SEQSCAN --- A(JOIN)
   end
```
Merging Projection and Selection into Get can be beneficial because storage engine may optimize the reading performance given the predicate. 
But for hash join and lateral join, I am not sure how knowing the projection columns and predicates can benefit.

Reply:
Option one does not make sense. Seqscan is always done on a table. I understand we have a separate node for storage but by definition, it should be part of seqscan.
 
Right now we are using option 2.

Reply:
I thought about it. To make things modular we can move predicate and projection logic to a util function. What do you think? I hope that should be fine with the ray right?

Reply:
For ray, it prefers option 2, since it has more options to insert the exchange operator for different parralision configuration. PROJECTION and SELECTOIN needs to be different operator nodes. They can not be within the JOIN exectutor. In other words, JOIN needs to only do the join job. Otherwise we can not inject the exchange operator.  
Regarding SEQSCAN, I am a bit confused. Is SEQSCAN === PROJECTION + SELECTION?

Reply:
Sounds good.
SEQSCAN **not equal** PROJECTION + SELECTION. It means iterating over the data. Ideally, our storage node + our seqscan node == true SeqScan